indexer/editable_map_object.cpp

```diff
@@ -70,6 +72,105 @@ bool IsProtocolSpecified(string const & website)
 {
   return GetProtocolNameLength(website) > 0;
 }
+osm::FakeNames MakeFakeSource(StringUtf8Multilang const & source,
+                              vector<int8_t> const & mwmLanguages, StringUtf8Multilang & fakeSource)
+{
+  string defaultName;
+  // Fake names works for mono language (official) speaking countries-only.
+  if (mwmLanguages.size() != 1 || !source.GetString(StringUtf8Multilang::kDefaultCode, defaultName))
+  {
+    return {};
+  }
+
+  osm::FakeNames fakeNames;
+  // Mwm name has higher priority then English name.
+  array<int8_t, kFakeNamesCount> fillCandidates = {{mwmLanguages.front(), StringUtf8Multilang::kEnglishCode}};
+  fakeSource = source;
+
+  string tempName;
+  for (auto const code : fillCandidates)
+  {
+    if (!source.GetString(code, tempName))
+    {
+      tempName = defaultName;
+      fakeSource.AddString(code, defaultName);
+    }
+
+    fakeNames.names.push_back({code, tempName});
+  }
+
+  if (!fakeNames.names.empty())
+    fakeNames.defaultName = defaultName;
+
+  return fakeNames;
+}
+
+// Tries to set default name from the localized name. Returns false when there's no such localized name.
+bool TryToFillDefaultNameFromCode(int8_t const code, StringUtf8Multilang & names)
+{
+  string newDefaultName;
+  if (code != StringUtf8Multilang::kUnsupportedLanguageCode)
+    names.GetString(code, newDefaultName);
+
+  // New default name can not be empty.
+  if (!newDefaultName.empty())
+  {
+    names.AddString(StringUtf8Multilang::kDefaultCode, newDefaultName);
+    return true;
+  }
+
+  return false;
+}
+
+// Tries to set default name to any non-empty localized name.
+// This is the case when fake names were cleared.
+void TryToFillDefaultNameFromAnyLanguage(StringUtf8Multilang & names)
+{
+  names.ForEach([&names](int8_t langCode, string const & name)
+  {
+    if (name.empty() || langCode == StringUtf8Multilang::kDefaultCode)
+      return true;
+
+    names.AddString(StringUtf8Multilang::kDefaultCode, name);
+    return false;
+  });
+}
+
+void RemoveFakesFromName(osm::FakeNames const & fakeNames, StringUtf8Multilang & name)
+{
+  vector<int8_t> codesToExclude;
+  string defaultName;
+  name.GetString(StringUtf8Multilang::kDefaultCode, defaultName);
+
+  for (auto const & item : fakeNames.names)
+  {
+    string tempName;
+    if (!name.GetString(item.m_code, tempName))
+      continue;
+    // No need to save in case when name is empty, duplicate of default name or was not changed.
+    if (tempName.empty() || tempName == defaultName ||
+        (tempName == item.m_filledName && tempName == fakeNames.defaultName))
+    {
+      codesToExclude.push_back(item.m_code);
+    }
+  }
+
+  if (codesToExclude.empty())
+    return;
+
+  StringUtf8Multilang nameWithoutFakes;
+  name.ForEach([&codesToExclude, &nameWithoutFakes](int8_t langCode, string const & value)
+  {
+    auto const it = find(codesToExclude.begin(), codesToExclude.end(), langCode);
+    if (it != codesToExclude.end())
```

Comment:
if (it == codesToExclude.end())
  nameWithoutFakes.AddString(langCode, value);
return true;

??